lib/provable-generic.ts

```diff
@@ -461,28 +513,35 @@ type InferPrimitiveJson<P extends Primitive> = P extends typeof String
   ? null
   : P extends undefined
   ? null
-  : JSONValue;
+  : any;
 
 type NestedProvable<Field> =
   | Primitive
+  | { provable: GenericProvable<any, any, Field> }
   | GenericProvable<any, any, Field>
   | [NestedProvable<Field>, ...NestedProvable<Field>[]]
   | NestedProvable<Field>[]
   | { [key: string]: NestedProvable<Field> };
 
-type InferProvable<A, Field> = A extends Constructor<infer U>
+type InferProvable<A, Field> = A extends { provable: Constructor<infer U> }
```

Comment:
nothing better than a nested conditional type :D imagine if TS would allow `if`, `else` and brackets here